File: manager/app/src/main/java/com/rifsxd/ksunext/ThRE.java
package id.thre.term;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.nio.charset.StandardCharsets;
import android.os.Message;

public class ThRE extends Thread implements Runnable {
	private volatile boolean done = false;
	private final Term term;
	private final Process process;
	private final OutputStream stdIn;
	private final BufferedReader stdOut;
	private Thread outReader;

	public ThRE(Term t3rm) {
		term = t3rm;
		Process p = null;
		OutputStream in = null;
		BufferedReader out = null;
		try {
			ProcessBuilder pb = new ProcessBuilder("/system/bin/sh");
			pb.redirectErrorStream(true); // merge stderr into stdout
			p = pb.start();
			in = p.getOutputStream();
			out = new BufferedReader(new InputStreamReader(p.getInputStream(), StandardCharsets.UTF_8));
		} catch (IOException e) {
			e.printStackTrace();
		}
		process = p;
		stdIn = in;
		stdOut = out;
	}

	private void startReaders() {
		if (stdOut != null) {
			outReader = new Thread(new Runnable() {
					@Override
					public void run() {
						readLoop(stdOut);
					}
				});
			outReader.start();
		}
	}

	private void readLoop(BufferedReader reader) {
		try {
			int ch;
			StringBuilder sb = new StringBuilder();
			while (!done && (ch = reader.read()) != -1) {
				sb.append((char) ch);
				String current = sb.toString();
				if (current.contains("[DONE]")) {
					String outputBeforeDone = current.substring(0, current.indexOf("[DONE]"));
					if (!outputBeforeDone.isEmpty()) {
						postToHandler(outputBeforeDone);
					}
					term.showNextPrompt();
					sb.setLength(0);
				} else if (ch == '\n') {
					postToHandler(current);
					sb.setLength(0);
				}
			}
		} catch (IOException e) {
			e.printStackTrace();
		}
	}

	private void postToHandler(final String text) {
		if (term == null || term.handle == null) return;
		term.handle.post(new Runnable() {
				@Override
				public void run() {
					String cleanText = text.replaceAll("\\u001b\\[[;\\d]*[A-Za-z]", "");
					term.getUI().output.append(cleanText);

				}
			});
	}

	@Override
	public void run() {
		startReaders();
		try {
			if (process != null) {
				process.waitFor();
			}
		} catch (InterruptedException e) {
			Thread.currentThread().interrupt();
		} finally {
			shutdown();
		}
	}

	public synchronized void exec(String command) {
		try {
			if (stdIn != null && process != null) {
				String[] lines = command.split("\n");
				for (String line : lines) {
					if (line.trim().isEmpty()) continue;
					stdIn.write((line + "\n").getBytes(StandardCharsets.UTF_8));
					stdIn.flush();
					try { Thread.sleep(50); } catch (InterruptedException e) {}
				}
			}
		} catch (IOException e) {
			e.printStackTrace();
		}
	}

	public synchronized void shutdown() {
		done = true;
		try {
			if (stdIn != null) stdIn.close();
		} catch (IOException ignored) {}
		try {
			if (stdOut != null) stdOut.close();
		} catch (IOException ignored) {}
		if (process != null) {
			process.destroy();
		}
		if (outReader != null) {
			outReader.interrupt();
		}
	}
}

File: manager/app/src/main/java/id/next/manager/ThRE.kt
package id.next.manager;

import android.os.Message
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.io.OutputStream
import java.nio.charset.StandardCharsets

class ThRE(private val term: Term) : Thread() {
 @Volatile
private var done = false
private val process: Process?
private val stdIn: OutputStream?
private val stdOut: BufferedReader?
private var outReader: Thread? = null

init {
    var p: Process? = null
    var inStream: OutputStream? = null
    var outReaderBuf: BufferedReader? = null
    try {
        val pb = ProcessBuilder("/system/bin/sh")
        pb.redirectErrorStream(true)
        p = pb.start()
        inStream = p.outputStream
        outReaderBuf = BufferedReader(InputStreamReader(p.inputStream, StandardCharsets.UTF_8))
    } catch (e: IOException) {
        e.printStackTrace()
    }
    process = p
    stdIn = inStream
    stdOut = outReaderBuf
    // readers started in run()
}

private fun startReaders() {
    stdOut?.let { reader ->
        outReader = Thread {
            readLoop(reader)
        }
        outReader?.start()
    }
}

private fun readLoop(reader: BufferedReader) {
    try {
        val sb = StringBuilder()
        while (!done) {
            val ch = reader.read()
            if (ch == -1) break
            sb.append(ch.toChar())
            // flush on newline or when buffer grows to prevent large delays
            if (ch == '\n'.code || sb.length >= 1024) {
                val chunk = sb.toString()
                sb.setLength(0)
                postToHandler(chunk)
            }
        }
        // flush remaining
        if (sb.isNotEmpty()) {
            postToHandler(sb.toString())
        }
    } catch (e: IOException) {
        e.printStackTrace()
    }
}

private fun postToHandler(text: String) {
    val handler = term.handle
    handler.post {
        val msg = Message.obtain(handler, 0, text)
        handler.sendMessage(msg)
    }
}

override fun run() {
    startReaders()
    try {
        process?.waitFor()
    } catch (e: InterruptedException) {
        Thread.currentThread().interrupt()
    } finally {
        shutdown()
    }
}

@Synchronized
fun exec(command: String) {
    try {
        if (stdIn != null && process != null && process.isAlive) {
            stdIn.write((command + "\n").toByteArray(StandardCharsets.UTF_8))
            stdIn.flush()
        }
    } catch (e: IOException) {
        e.printStackTrace()
    }
}

@Synchronized
fun shutdown() {
    done = true
    try {
        stdIn?.close()
    } catch (ignored: IOException) {}
    try {
        stdOut?.close()
    } catch (ignored: IOException) {}
    process?.destroy()
    outReader?.interrupt()
}

}